Zend/zend_partial.c

```diff
@@ -0,0 +1,589 @@
+    }
+    
+    efree(fci.params);
+}
+
+void zend_partial_apply(zval *result, uint32_t info, zend_execute_data *call) {
+    zend_function *function, *prototype = NULL;
+
+    ZVAL_OBJ(result, zend_partial_new(zend_ce_closure));
+
+    zend_partial *parent, *partial = (zend_partial*) Z_OBJ_P(result);
+
+    if ((parent = zend_partial_parent(&call->This))) {
+        function  = &parent->func;
+        prototype = &parent->prototype;
+
+        partial->argc = parent->argc + ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, parent->argv, parent->argc * sizeof(zval));
+
+        uint32_t arg = 0,
+                 aend = parent->argc;
+        zval *param = ZEND_CALL_ARG(call, 1),
+             *pend = param + ZEND_CALL_NUM_ARGS(call);
+
+        while (arg < aend && param < pend) {
+            if (Z_TYPE(partial->argv[arg]) == IS_UNDEF) {
+                if (!Z_ISUNDEF_P(param)) {
+                    ZVAL_COPY_VALUE(&partial->argv[arg], param);
+                }
+                param++;
+            }
+            arg++;
+        }
+
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            if (parent->named) {
+                partial->named = zend_array_dup(parent->named);
+
+                zend_hash_merge(partial->named, call->extra_named_params, zval_copy_ctor, true);
+            } else {
+                partial->named = call->extra_named_params;
+                GC_ADDREF(call->extra_named_params);
+            }
+        }
+    } else {
+        function = call->func;
+
+        partial->argc = ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, ZEND_CALL_ARG(call, 1), partial->argc * sizeof(zval));
+        
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            partial->named = call->extra_named_params;
+            GC_ADDREF(call->extra_named_params);
+        }
+    }
+
+    memcpy(&partial->func, function, (function->type == ZEND_INTERNAL_FUNCTION) ?
+              sizeof(zend_internal_function) : sizeof(zend_op_array));
+
+    partial->func.common.fn_flags &= ~ZEND_ACC_CLOSURE;
+    partial->func.common.fn_flags |= ZEND_ACC_IMMUTABLE;
+    
+    if (partial->func.common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE) {
+        partial->func.common.fn_flags |= ZEND_ACC_TRAMPOLINE_PERMANENT;
+    }
+
+    if (info & ZEND_APPLY_FACTORY) {
+        ZVAL_PTR(&partial->This, Z_OBJCE(call->This));
+
+        GC_ADD_FLAGS(Z_OBJ(call->This), IS_OBJ_DESTRUCTOR_CALLED);
+        OBJ_RELEASE(Z_OBJ(call->This));
+        
+        ZEND_ASSERT(partial->func.common.fn_flags & ZEND_ACC_CTOR);
+    } else {
+        ZVAL_COPY(&partial->This, &call->This);
+
+        if (info & ZEND_APPLY_PASS) {
+            partial->func.common.fn_flags |= ZEND_ACC_CTOR;
+        } else {
+            partial->func.common.fn_flags &= ~ZEND_ACC_CTOR;
```

Comment:
Why is this flag unset necessary? I would expect that the partial implementation would be checking for PTR This, rather than the ambiguous CTOR flag.

Reply:
The unset is so that directly applied constructors behave like normal functions. PTR This is used for factory application.